ProjectEureka/project-eureka-flutter
# PE-50 | Video Communication

UPDATE:
Backend was deployed to AWS, so use Eureka IP address instead of Localhost to test it in Postman, if you want. Flutter is already connected to AWS in the last commit, so you only need to add AGORA_ID to .env file. So you can ignore all instructions regarding Flutter and Spring Boot connection, as it now does not require your local IP.

**What I did:**

- Created a demo page with "call" and "answer" buttons
- Created services and controllers on Backend to generate token once calling, request generated token for user who answers the call, deleting the token once call is ended
- Created service file in Flutter to communicate with backend
- Created "call page" on which users can see each other's video, switch camera, mute audio, mute video, end hung up the call

**How Agora SDK works and how I implemented communication**

- Agora SDK was used as a video communication service in Eureka. Currently it allows a thousand hours per month free. In order to connects users in a call, Agora service on backend must receive at least a token and a channel name, both are of type Strings. Also, on Backend, agora will receive a certificate string. Both front end and back end use APP_ID which allows generated token to be recognized in frontend, through some encryption/decryption/
- Demo page is simulating buttons "Call" and "Answer" that will be later added to the chat page between two users
<img src="https://user-images.githubusercontent.com/68677129/114325772-ae28d800-9ae6-11eb-924b-bc3a8f3e76e4.jpg" width="300">

- When user clicks "Call" on demo page, flutter combines the caller user ID (which is current user id!) AND target user ID (another words - recipient) in a single string. For example: Caller user ID is "123ab" and target user ID is "789cd", then the combined string will be "123ab789cd". This will be the Channel name which will bring both users to the same channel. 
- After Flutter sends this string to backend (through a GET operation link), spring boot application will use service files to generate a token, and will save it into a HashMap, like the following: "123ab789cd"="some_long_generated_token". 
- After token is saved in HashMap, caller will receive the token and will be directed to "call page", on which user will see "Calling..." status, meaning that he is waiting for the target user to answer. Screen is black, but that will be explained later. 
<img src="https://user-images.githubusercontent.com/68677129/114325448-1080d900-9ae5-11eb-9a11-f4db4d65f8c1.jpg" width="300">

- On another end, target user (recipient) clicks "Answer". This will create a similar string of the channel name, just like for the caller. However, in this case it will be a combination of current user ID, which is the target user, and the caller ID, which is the the caller. So the combination will also be"123ab789cd". This string will be sent to backend through a different GET operation link, which will search for token associated with the given string in HashMap
- If the call wasn't started by the companion or the call was ended, user who clicks "Answer" will see an alert message "Call is inactive", since backend will return an error, as the token wasn't found. Once video communication is integrated to the chat page, answer button will appear once user receives the call, and if call was ended, user cannot answer again and go back to the call.
![Screenshot_2](https://user-images.githubusercontent.com/68677129/114325517-59389200-9ae5-11eb-961c-419e2926bc3a.jpg)

- If call was started, and recipient clicked "answer", users will be connected, as they joined the same channel and both used the same token. Screenshot below was taken from emulator. On top of the screenshot is the camera of the physical phone that was used to test video communication , on the bottom is the emulator's simulated camera.
<img src="https://user-images.githubusercontent.com/68677129/114326040-c3523680-9ae7-11eb-9613-ced43d62f909.jpg" width="300">

- Mute audio and switch camera buttons work as intended, however, mute video button works like a "pause" button, meaning that video will not disappear, but will be paused.
- If any of the users will hung up the call, the DELETE operation will be sent to backend and the token will be remove from HashMap. Also, both users will have a function triggered to make both ends to finish the call, and there be the following message displayed
![Screenshot_5](https://user-images.githubusercontent.com/68677129/114326245-c1d53e00-9ae8-11eb-8613-884ceb7325a4.jpg)

- If any of the user click "back" button, it will ask them if they want to leave. If user clicks "leave", it will work just like end call button, thus, both users will be kicked from the call.
![Screenshot_6](https://user-images.githubusercontent.com/68677129/114326319-0bbe2400-9ae9-11eb-8f78-da484c19abb5.jpg)

**How to test:**

- Checkout to PE-80 branch in Spring Boot application: https://github.com/ProjectEureka/project-eureka-spring/pull/12
- Add two environmental variables to "application.properties" file. You can find those on Discord or Google Drive
- Run `mvn spring-boot:run` in console from the root directory
- Open postman and test the following operations (NOTE: on screenshots are outdated links, they must be in a format "v1/agora/...")
  - GET on `http://localhost:8080/v1/agora/call/<enter here any string here>`. This must return the generated token
![Screenshot_8](https://user-images.githubusercontent.com/68677129/114326485-e2ea5e80-9ae9-11eb-9d3e-aa594f17dae9.jpg)

  - GET on `http://localhost:8080/v1/agora/answer/<enter here THE SAME string>`. This must return the same token as for the caller
![Screenshot_9](https://user-images.githubusercontent.com/68677129/114326506-feee0000-9ae9-11eb-9264-503b33787dcd.jpg)

  - If token wasn't found, it will return an error string instead as a token. This error string will be used by flutter in order to display "call is inactive" message
  - DELETE on `http://localhost:8080/v1/agora/hungup/<enter here any string>`. This must return a confirmation string that call was ended
![Screenshot_7](https://user-images.githubusercontent.com/68677129/114326571-4a081300-9aea-11eb-9d97-0a7959172d38.jpg)

- Now checkout to PE-50 (Flutter). Run `flutter pub get` to install dependencies
- Update .end file to add Agora APP ID. You can Find it on Discord or Google Drive
- Run application
- Depending on your emulator or physical phone, you may need to change the URL base in video_communication.dart service file (there are 3 URLs), which is a link to communicate with local Spring Boot. In my case, localhost wasn't recognized, so it was either 10.0.2.2 or 192.168.1.2, but in your case it may be different if it won't connect, keep it in mind.
- Go to side menu. There will be a "Video" option which brings to the video communication demo page
- Test the functionalities. You can also follow the messages printed by Spring Boot application to see if API calls are working properly
- If you can connect physical phone to Android Studio, you could test the video communication between two users, as well as different case scenarios
- Currently, target user ID is simulated, but once integrated in Chat system, it will work as intended

**Limitations and bugs:**
- Current Agora SDK for video communication does not allow to turn off the camera completely, so only Mute Video works.
- If user closes the application, function to hung up the call and remove token from backend is not gonna work. But as soon as user calls again, token will be overwritten. Perhaps, later it will be changed, and function to hung up will be triggered once the application is closed with an ongoing call.
- Once starting a call, video preview is not working, it will turn on once both users are connected to the call. There was no solution found so far to show video from camera before another user joins the channel
